lib/schedule/block.ex

```diff
@@ -1,101 +1,95 @@
-
-    nonrevenue_trips =
-      Enum.group_by(nonrevenue_trips, fn trip -> {trip.block_id, trip.schedule_id} end)
-
-    revenue_trips
-    |> Enum.filter(fn trip -> trip.stop_times != [] end)
-    |> Enum.group_by(fn trip -> {trip.block_id, trip.service_id} end)
-    |> Helpers.map_values(fn trips ->
-      block_id = List.first(trips).block_id
-      schedule_id = List.first(trips).schedule_id
-      deadheads = Map.get(nonrevenue_trips, {block_id, schedule_id}, [])
-      block_from_trips(trips, deadheads)
-    end)
+  @spec key(t()) :: key()
+  def key(block) do
+    {block.schedule_id, block.id}
+  end
+
+  @spec blocks_from_pieces([Piece.t()]) :: by_id()
+  def blocks_from_pieces(pieces) do
+    pieces
+    |> Enum.reject(&Piece.from_non_current_rating?/1)
+    |> Enum.group_by(&Piece.block_key/1)
+    |> Helpers.map_values(&block_from_pieces(&1))
   end
 
-  @spec block_from_trips([Trip.t()], [Trip.t()]) :: t()
-  def block_from_trips(revenue_trips, nonrevenue_trips \\ []) do
-    revenue_trips = Enum.sort_by(revenue_trips, & &1.start_time)
-    nonrevenue_trips = Enum.sort_by(nonrevenue_trips, & &1.start_time)
-    first_trip = List.first(revenue_trips)
-
-    start_time =
-      if nonrevenue_trips == [] do
-        first_trip.start_time
-      else
-        min(
-          first_trip.start_time,
-          List.first(nonrevenue_trips).start_time
-        )
-      end
-
-    end_time =
-      if nonrevenue_trips == [] do
-        List.last(revenue_trips).end_time
-      else
-        max(
-          List.last(revenue_trips).end_time,
-          List.last(nonrevenue_trips).end_time
-        )
-      end
+  @spec block_from_pieces([Piece.t()]) :: t()
+  def block_from_pieces(pieces) do
+    pieces = Enum.sort_by(pieces, & &1.start_time)
+    first_piece = List.first(pieces)
+    last_piece = List.last(pieces)
+
+    first_revenue_trip =
+      pieces
+      |> Enum.flat_map(& &1.trips)
+      |> Enum.reject(&match?(%AsDirected{}, &1))
+      |> Enum.find(&Trip.is_revenue_trip?/1)
 
     %__MODULE__{
-      id: first_trip.block_id,
-      service_id: first_trip.service_id,
-      schedule_id: first_trip.schedule_id,
-      start_time: start_time,
-      end_time: end_time,
-      trips: revenue_trips
+      id: first_piece.block_id,
+      service_id: first_revenue_trip.service_id,
+      schedule_id: first_piece.schedule_id,
+      start_time: first_piece.start_time,
+      end_time: last_piece.end_time,
+      pieces: pieces
     }
   end
 
-  @spec get(by_id(), id(), Service.id()) :: t() | nil
-  def get(by_id, block_id, service_id) do
-    by_id[{block_id, service_id}]
+  @spec revenue_trips(t()) :: [Trip.t()]
+  def revenue_trips(%__MODULE__{pieces: pieces}) do
+    pieces
+    |> Enum.flat_map(& &1.trips)
+    |> Enum.reject(&match?(%AsDirected{}, &1))
+    |> Enum.filter(&Trip.is_revenue_trip?/1)
+  end
+
+  @spec get(by_id(), id(), Hastus.Schedule.id()) :: t() | nil
+  def get(by_id, schedule_id, block_idd) do
```

Comment:
Is `block_idd` intentional or a typo? Also the arguments and on this line and in the typespec on the previous line seem to be in the opposite order.

Reply:
Typo. And this is why I can't wait for Gleam to be production-ready, will fix the spec.